apps/email/js/cards/cmp/autocomplete.js

```diff
@@ -0,0 +1,565 @@
+      document.addEventListener('click', this.onDocumentEvent, true);
+      document.addEventListener('focus', this.onDocumentEvent, true);
+    },
+
+    /**
+     * Custom element callback when element is removed from the document.
+     */
+    detachedCallback: function() {
+      window.removeEventListener('resize', this.onResize, false);
+      document.removeEventListener('click', this.onDocumentClick, true);
+      document.addEventListener('focus', this.onDocumentEvent, true);
+    },
+
+    /**
+     * Triggered by input focus events.
+     */
+    onFocus: function(event) {
+      this.inputNode = event.target;
+
+      // If existing label node, be sure to remove the active class from it. In
+      // cases where the focus goes from one input to the other, hide() does not
+      // run to clear this state.
+      if (this.inputLabelNode) {
+        this.inputLabelNode.classList.remove('cmp-autocomplete-active');
+      }
+
+      this.inputLabelNode = findAncestorWithClass(this.inputNode,
+                                                'cmp-autocomplete-input-label');
+      this.inputLabelNode.classList.add('cmp-autocomplete-active');
+
+      this.query = this.inputNode.value.trim();
+      this.runThrottledQuery();
+    },
+
+    /**
+     * Triggered by any click or focus in the document. Used to determine if the
+     * autocomplete should be hidden.
+     */
+    onDocumentEvent: function(event) {
+      var target = event.target;
+
+      // If no classList, then it is not an element, and should be discarded.
+      if (!target.classList) {
+        this.hide();
+        return;
+      }
+
+      // If an autocomplete input, do not do anything, handled by other code in
+      // this module.
+      if (target.classList.contains('cmp-autocomplete-input')) {
+        return;
+      }
+
+      // Also ignore if it is part of an autocomplete-item.
+      if (getAutocompleteItem(target)) {
+        return;
+      }
+      this.hide();
+    },
+
+    onKeyDown: function(event) {
+      // If the down arrow is pressed, then focus on the first autocomplete
+      // item, if there is one.
+      if (event.key === 'ArrowDown' && !this.isHidden) {
+        var node = this.list.querySelector('.cmp-autocomplete-item-link');
+        if (node) {
+          node.focus();
+          event.stopPropagation();
+          event.preventDefault();
+        }
+      }
+    },
+
+    /**
+     * Triggered by input element input events.
+     */
+    onInput: function(event) {
+      this.query = this.inputNode.value.trim();
+      this.runThrottledQuery();
+    },
+
+    /**
+     * Handles window resizes. Could be triggered by keyboard size changes.
+     */
+    onResize: function(event) {
+      if (this.inputNode && !this.isHidden) {
+        this.positionElements();
+      }
+    },
+
+    onMatchKeyDown: function(event) {
+      if (event.key === 'Enter' && !this.isHidden) {
+        this.onMatchClick(event);
+      }
```

Comment:
Agreed it is weird, and feels like for it all to work properly, I need to hide more of the the rest of the DOM, either via tabindex=-1 or some other trickery for tab to just work. Probably need to explicitly set tab order so the autocomplete matches appear between valid inputs: if in "to:", tab should be from "to:", all autocomplete entries, then to "cc:".

I do not like the idea of all of that manual setting and unsetting though. For now I just removed this arrowdown support. If you would like to see it with this bunch, then I'll do a separate commit pass in this pull request for it.